From f5946221504c14ea056818ffd609f7aa5156da4f Mon Sep 17 00:00:00 2001
From: Gerald Combs <gerald@wireshark.org>
Date: Mon, 11 Apr 2016 15:38:14 -0700
Subject: [PATCH 1/6] Restrict the depth of the protocol tree.

Limit the the protocol tree to 500 levels. This keeps us from running
out of stack space in proto_tree_traverse_{pre,post}_order.

Bug: 12268
Change-Id: I60f6b4487bf51a4d43da0429819a8a20fbdb313a
Reviewed-on: https://code.wireshark.org/review/14887
Petri-Dish: Gerald Combs <gerald@wireshark.org>
Tested-by: Petri Dish Buildbot <buildbot-no-reply@wireshark.org>
Reviewed-by: Gerald Combs <gerald@wireshark.org>
(cherry picked from commit 8dc9551e1d56290e6f7f02cc38b77e1d211fd4a5)
Reviewed-on: https://code.wireshark.org/review/15010
(cherry picked from commit c5a9678f5f601a86745e5108323a573171d23fac)
Reviewed-on: https://code.wireshark.org/review/15442
Reviewed-by: Balint Reczey <balint@balintreczey.hu>
---
 epan/proto.c | 25 +++++++++++++++++++++++--
 1 file changed, 23 insertions(+), 2 deletions(-)

diff --git a/epan/proto.c b/epan/proto.c
index a1a34a4..2f8d387 100644
--- a/epan/proto.c
+++ b/epan/proto.c
@@ -59,11 +59,14 @@
 
 #include <wsutil/plugins.h>
 
+/* Ptvcursor limits */
 #define SUBTREE_ONCE_ALLOCATION_NUMBER 8
 #define SUBTREE_MAX_LEVELS 256
-/* Throw an exception if we exceed this many tree items. */
-/* XXX - This should probably be a preference */
+
+/* Throw an exception if our tree exceeds these. */
+/* XXX - These should probably be preferences */
 #define MAX_TREE_ITEMS (1 * 1000 * 1000)
+#define MAX_TREE_LEVELS (5 * 100)
 
 
 typedef struct __subtree_lvl {
@@ -3757,6 +3760,24 @@ proto_tree_add_node(proto_tree *tree, field_info *fi)
 {
 	proto_node *pnode, *tnode, *sibling;
 	field_info *tfi;
+	int depth = 1;
+
+	/*
+	 * Restrict our depth. proto_tree_traverse_pre_order and
+	 * proto_tree_traverse_post_order (and possibly others) are recursive
+	 * so we need to be mindful of our stack size.
+	 */
+	if (tree->first_child == NULL) {
+		for (tnode = tree; tnode != NULL; tnode = tnode->parent) {
+			depth++;
+			if (G_UNLIKELY(depth > MAX_TREE_LEVELS)) {
+				THROW_MESSAGE(DissectorError, wmem_strdup_printf(wmem_packet_scope(),
+						     "Maximum tree depth %d exceeded for \"%s\" - \"%s\" (%s:%u)",
+						     MAX_TREE_LEVELS,
+						     fi->hfinfo->name, fi->hfinfo->abbrev, G_STRFUNC, __LINE__));
+			}
+		}
+	}
 
 	/*
 	 * Make sure "tree" is ready to have subtrees under it, by
-- 
2.1.4

